acra-censor/handlers/querycapture_handler.go

```diff
@@ -38,99 +37,59 @@ type QueryInfo struct {
-	var _, err = os.Stat(filePath)
-
-	// create file if not exists
-	if err != nil {
-		if os.IsNotExist(err) {
-			var file, err = os.Create(filePath)
-			if err != nil {
-				return nil, err
-			}
-			defer file.Close()
-		} else {
-			return nil, err
-		}
-	}
-
-	bufferBytes, err := ioutil.ReadFile(filePath)
+	// open or create file, APPEND MODE
+	openedFile, err := os.OpenFile(filePath, os.O_APPEND|os.O_RDWR|os.O_CREATE, 0600)
 	if err != nil {
+		log.WithError(ErrSingleQueryCaptureError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorSecurityError)
 		return nil, err
 	}
 
-	var queries []QueryInfo
-
-	if len(bufferBytes) != 0 {
-		if err = json.Unmarshal(bufferBytes, &queries); err != nil {
-			return nil, err
-		}
-	}
-
-	logChannel := make(chan QueryInfo, MaxQueriesInChannel)
-
+	// signals
 	signalShutdown := make(chan os.Signal, 2)
 	signal.Notify(signalShutdown, os.Interrupt, syscall.SIGTERM)
 
 	signalBackgroundExit := make(chan bool)
 
+	// create handler
 	handler := &QueryCaptureHandler{}
 	handler.filePath = filePath
-	handler.Queries = queries
-	handler.logChannel = logChannel
 	handler.signalBackgroundExit = signalBackgroundExit
 	handler.serializationTimeout = DefaultSerializationTimeout
 	handler.serializationTicker = time.NewTicker(DefaultSerializationTimeout)
 
-	f, err := os.OpenFile(filePath, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0600)
-	if err != nil {
-		log.WithError(ErrSingleQueryCaptureError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorSecurityError)
-		return nil, err
-	}
+	// read existing queries from file
+	handler.ReadAllQueriesFromFile()
+
 
 	//handling goroutine
 	go func() {
 		for {
 			select {
 			case <-handler.serializationTicker.C:
-				err := handler.Serialize()
+				err := handler.DumpBufferedQueriesToFile(openedFile)
 				if err != nil {
 					log.WithError(ErrComplexSerializationError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorSecurityError)
 				}
 				handler.serializationTicker.Stop()
 				handler.serializationTicker = time.NewTicker(handler.serializationTimeout)
 
-			case queryInfo, ok := <-handler.logChannel:
-				if ok {
-					bytes, err := json.Marshal(queryInfo)
-					if err != nil {
-						log.WithError(ErrSingleQueryCaptureError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorSecurityError)
-					}
-
-					if _, err = f.WriteString("\n"); err != nil {
-						log.WithError(ErrSingleQueryCaptureError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorSecurityError)
-					}
-
-					if _, err = f.Write(bytes); err != nil {
-						log.WithError(ErrSingleQueryCaptureError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorSecurityError)
-					}
-
-				} else {
-					//channel is unexpectedly closed
-					log.WithError(ErrUnexpectedCaptureChannelClose).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorSecurityError)
-				}
-
+			// TODO: how to remove duplicate code?
```

Comment:
Golang doesn't provide way to write two cases at once like in switch with `fallthrough` becase any read from channel block goroutine and for `fallthrough` it need to read second channel :) So we can hack it like:
1. move logic to some function
```
func (handler *QueryCaptureHandler) finish(){
  handler.serializationTicker.Stop()
  err := handler.DumpAllQueriesToFile()
  if err != nil {
      log.WithError(ErrComplexSerializationError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorSecurityError)
  }
  openedFile.Close()
}
```
and call it in two cases
```
case <-someCh:
....
   handler.finish()
   return
....
```

2. use one more channel that will flag that we need to finish
```
exitCh := make(chan bool)
for {
  select {
     case <-signalBackgroundExit:
       exitCh <- true
     case <-signalShutdown:
       exitCh <- true
  }
  select {
    case <-exitCh:
       // exit code
    default:
      // do nothing, continue
  }
}
```